### PR TITLE
#170947666 make one travel request has multiple stops 

### PR DESCRIPTION
**Pull request by EugeneMunya**
#### What does this PR do?
- Add functionality for letting requester to make one travel request that has multiple destinations embedded.
#### Description of Task to be completed?
- As requester I will be able to request to travel in more that one destination at once without creating separate request for each destination.
#### How should this be manually tested?
- open your terminal and run this command `git clone https://github.com/andela/knightsbn    backend.git` to clone repository
- run `cd knights-bn-backend` to get inside the project
- run `git fetch && git checkout ft-make-multi-city-request-170947666` 
- run `npm install` to install dependencies
- check if you have PostgresSQL database if you don't have it install it from here [PostgreSQL](https://www.2ndquadrant.com/en/blog/pginstaller-install-postgresql/)
- create database
- create `.env` file in your project root directory
- follow `env.example` file to create environment variables in `.env` file and 
   assign them with all your database configurations
- run `npm run createTables` to create tables in your new database you have created.
- run `npm run test` to run all the tests
- run `npm run dev-start` to start the server.
- open your postman or any other API tester you have
-  to create your request you must be logged in or signed up , see how to do [log in](https://github.com/andela/knights-bn-backend/pull/23) or [sign up](https://github.com/andela/knights-bn-backend/pull/17)
- to test multi-city request send POST request to this link `http://localhost:4000/api/v1/trips/request/multicity` with the data object defined like the one below in your postman body.
```
{
    "origin": "New york",
    "destination": "East Africa",
    "departureDate": "2020-04-28",
    "returnDate": "2022-09-30",
   "accommodation": "XYZ campus",
    "type": "multi_way",
    "reason": "Reaseach",
    "passportNumber": "12345677",
    "cities": [
      {
        "name": "kigali",
        "from": "2020-05-02",
        "to": "2020-08-31"
      },
       {
        "name": "Nairob",
        "from": "2020-06-02",
        "to": "2020-09-08"
      }
    ]
  }
```
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#170947666](https://www.pivotaltracker.com/story/show/170947666)
#### Screenshots

Screen Shot 2020-03-03 at 09 02 56

### CODE REVIEW

**Comment by Niyonsengaeric** [comment]
Screen Shot 2020-02-28 at 16 42 32
you should save also the requesterId

**Comment by Niyonsengaeric** [comment]
I think you can also not allow users to make the same request more than one.

**Comment by EugeneMunya** [comment]
> I think you can also not allow users to make the same request more than one.

I am still working on it I will be done in few

**Comment by EugeneMunya** [comment]
> Screen Shot 2020-02-28 at 16 42 32
> 
> you should save also the requesterId

Let  me solve it.

**Comment by Niyonsengaeric** [comment]
Screen Shot 2020-02-28 at 17 04 09
you are also missing some date validation check on the dates  that am using on the above picture

**Comment by EugeneMunya** [comment]
> I think you can also not allow users to make the same request more than one.

I have managed to implement this feedback you can for now you can check how it works

**Comment by EugeneMunya** [comment]
> Screen Shot 2020-02-28 at 17 04 09
> 
> you are also missing some date validation check on the dates that am using on the above picture

I tried to validation the date you can take a look

**Comment by EugeneMunya** [comment]
> Screen Shot 2020-02-28 at 17 04 09
> 
> you are also missing some date validation check on the dates that am using on the above picture

**Comment by EugeneMunya** [comment]
> > Screen Shot 2020-02-28 at 17 04 09
> > you are also missing some date validation check on the dates that am using on the above picture
> 
> I tried to validate the date  for now  you can take a look.

**Comment by EugeneMunya** [comment]
> > Screen Shot 2020-02-28 at 16 42 32
> > you should save also the requesterId
> 
> Let me solve it.

I have managed to do it, now you can re-test again and see

**Comment by EugeneMunya** [comment]
> Screen Shot 2020-02-28 at 17 04 09
> 
> you are also missing some date validation check on the dates that am using on the above picture

I managed to validate them but it was not easy task

**Comment by Niyonsengaeric** [comment]
Screen Shot 2020-02-29 at 19 49 25
@EugeneMunya  now it is not working on my side.

**Comment by EugeneMunya** [comment]
> Screen Shot 2020-02-29 at 19 49 25
> 
> @EugeneMunya now it is not working on my side.

@Niyonsengaeric I have resolved the issue you can check for now

**Review by Niyonsengaeric — APPROVED** [review]

File: src/controllers/request.js [hunk 1]
```diff
@@ -95,4 +94,22 @@ export default class usersController {
...
+    const {lineManager} = requester;
+    const manager = await models.User.findOne({where:{email:lineManager}});
+    if(manager === 'null') return res.status(404).json({status:404, error:'It seems your manager is not avilable in the system'})
+    req.body.requesterId = req.user.id;
+    req.body.managerId = manager.id;
+    req.body.status = 'pending';
+    const multiCity = await models.Request.create(req.body);
+    return res.status(201).json({status:201, message:'Your request has successfully created', data:multiCity});
+   }catch(error){
+    return res.status(500).json({status:500, err: 'Database Error'});
```
Comment (Kagorora) [inline src/controllers/request.js]: @EugeneMunya  Please catch the actual error i.e you have caught the error but you have hard copied your own error ('Database Error'), please throw the actual error you have caught.

File: src/controllers/userProfile.js [hunk 2]
```diff
@@ -13,7 +16,7 @@ export default class changeUserProfile {
 
   static changeMyProfileInfo(request, response) {
     request.body.birthDay = validateDate(request.body.birthDay);
-    request.body.profileImage = (typeof request.file === 'undefined') ? null : `http://localhost:4000/${request.file.path}`;
+    request.body.profileImage = (typeof request.file === 'undefined') ? null : `${process.env.URL}/${request.file.path}`;
```
Comment (Kagorora) [inline src/controllers/userProfile.js]: @EugeneMunya  Please change the  name of the variable from URL to LOCAL_HOST

File: src/tests/multicityRequest.test.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        });
+    });
+
+    it('should return 200 on successful created request ', (done) => {
+      chai
+        .request(app)
+        .post('/api/v1/trips/request/multicity')
+        .send(mockData.multiCityRequest)
+        .end((err, res) => {
+          expect(res.status).to.equal(201);
```
Comment (Kagorora) [inline src/tests/multicityRequest.test.js:30]: @EugeneMunya  please include the message tests as well, this applies on the rest of the tests

File: src/controllers/request.js (hunk 1 above)
Comment (EugeneMunya) [reply src/controllers/request.js]: I have corrected it now you can check

File: src/controllers/userProfile.js (hunk 2 above)
Comment (EugeneMunya) [reply src/controllers/userProfile.js]: I have corrected that mistake for now you can check

File: src/tests/multicityRequest.test.js (hunk 3 above)
Comment (EugeneMunya) [reply src/tests/multicityRequest.test.js:30]: I tested also the messages, you see how I managed to do it

**Review by kagabof — APPROVED** [review]
Great job.

**Pull request merged**